src/scanner.c

```diff
@@ -0,0 +1,442 @@
+#include "tag.h"
+#include <wctype.h>
+
+enum TokenType {
+    START_TAG_NAME,
+    SCRIPT_START_TAG_NAME,
+    STYLE_START_TAG_NAME,
+    END_TAG_NAME,
+    ERRONEOUS_END_TAG_NAME,
+    SELF_CLOSING_TAG_DELIMITER,
+    IMPLICIT_END_TAG,
+    RAW_TEXT,
+    COMMENT
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    Tag *data;
+} tags_vec;
+
+typedef struct {
+    tags_vec tags;
+} Scanner;
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define VEC_RESIZE(vec, _cap)                                                  \
+    void *tmp = realloc((vec).data, (_cap) * sizeof((vec).data[0]));           \
+    assert(tmp != NULL);                                                       \
+    (vec).data = tmp;                                                          \
+    (vec).cap = (_cap);
+
+#define VEC_GROW(vec, _cap)                                                    \
+    if ((vec).cap < (_cap)) {                                                  \
+        VEC_RESIZE((vec), (_cap));                                             \
+    }
+
+#define VEC_PUSH(vec, el)                                                      \
+    if ((vec).cap == (vec).len) {                                              \
+        VEC_RESIZE((vec), MAX(16, (vec).len * 2));                             \
+    }                                                                          \
+    (vec).data[(vec).len++] = (el);
+
+#define VEC_POP(vec)                                                           \
+    {                                                                          \
+        STRING_FREE(VEC_BACK(vec).custom_tag_name);                            \
+        (vec).len--;                                                           \
+    }
+
+#define VEC_BACK(vec) ((vec).data[(vec).len - 1])
+
+#define VEC_FREE(vec)                                                          \
+    {                                                                          \
+        if ((vec).data != NULL)                                                \
+            free((vec).data);                                                  \
+        (vec).data = NULL;                                                     \
+    }
+
+#define VEC_CLEAR(vec)                                                         \
+    {                                                                          \
+        for (int i = 0; i < (vec).len; i++) {                                  \
+            STRING_FREE((vec).data[i].custom_tag_name);                        \
+        }                                                                      \
+        (vec).len = 0;                                                         \
+    }
+
+#define STRING_RESIZE(vec, _cap)                                               \
```

Comment:
same, we could factorize in a a string.h or maybe string_basic.h to avoid confusion with string.h